wallet-api/src/Language/PlutusTx/Interval.hs

```diff
@@ -0,0 +1,149 @@
+    deriving newtype (Num, Real, Integral, Serialise)
+
+makeLift ''Slot
+
+-- | A possibly open-ended inclusive-exclusive interval of slots.
+data Interval a = Interval { ivFrom :: Maybe a, ivTo :: Maybe a }
+    deriving (Eq, Ord, Show)
+    deriving stock (Generic)
+    deriving anyclass (ToSchema, FromJSON, ToJSON, Serialise)
+
+makeLift ''Interval
+
+type SlotRange = Interval Slot
+
+{- note [Definition of Interval]
+
+The purpose of this module is to provide an interval data type that 
+can be used in on-chain and off-chain code alike. Its two main uses in the 
+'wallet-api' module are the validity range of transactions, and the ranges of
+funds and slot numbers for wallet triggers.
+
+To ensure that 'Interval' can be used in on-chain code, the functions in this 
+module do not use type classes or functions from the Haskell Prelude. The types 
+of the query functions exported from here are specialised to
+'Interval Slot'. 
+
+-}
+
+
+-- | A 'SlotRange' that covers every slot.
+always :: Q (TExp SlotRange)
+always = [|| Interval Nothing Nothing ||]
+
+-- | An open-ended 'Interval' starting
+from :: Q (TExp (a -> Interval a))
+from = [|| \s -> Interval (Just s) Nothing ||]
+
+-- | An 'Interval' from negative infinity to a value (exclusive)
+to :: Q (TExp (a -> Interval a))
+to = [|| \s -> Interval Nothing (Just s) ||]
+
+-- | An inclusive-exclusive 'Interval' of 'a's.
+interval :: Q (TExp (a -> a -> Interval a))
+interval = [|| \s s' -> Interval (Just s) (Just s') ||]
+
+-- | A 'SlotRange' that covers only a single slot.
+singleton :: Q (TExp (Slot -> SlotRange))
+singleton = [|| \(Slot s) -> Interval (Just (Slot s)) (Just (Slot (s + 1))) ||]
+
+-- | Check if a 'SlotRange' is empty
+empty :: Q (TExp (SlotRange -> Bool))
+empty = [|| 
+    \(Interval f t) -> 
+        case f of 
+            Nothing -> False
+            Just f' -> case t of
+                Nothing -> False
+                Just t' -> f' >= t'
+    ||]
+
+-- | Check if 'Slot' is contained in a 'SlotRange'.
+member :: Q (TExp (Slot -> SlotRange -> Bool))
+member = [|| 
+    \(Slot s) (Interval f t) -> 
+        let lw = case f of { Nothing -> True; Just (Slot f') -> f' <= s; }
+            hg = case t of { Nothing -> True; Just (Slot t') -> t' > s; }
+        in
+            if lw then hg else False
+    ||]
+
+-- | Number of 'Slot's covered by the interval. @width (from x) == Nothing@
+width :: Q (TExp (SlotRange -> Maybe Int))
+width = [|| 
+    \(Interval f t) -> 
+        case f of 
+            Nothing -> Nothing
+            Just (Slot f') -> case t of
+                Nothing -> Nothing
+                Just (Slot t') -> Just (t' - f')  ||]
+
+-- | @a `contains` b@ is true if the 'SlotRange' @b@ is entirely contained in 
+--   @a@. That is, @a `contains` b@ if for every slot @s@, if @member s b@ then
+--   @member s a@.
+contains :: Q (TExp (SlotRange -> SlotRange -> Bool))
+contains = [|| 
+    \(Interval af at) (Interval bf bt) ->
+        let lw = case af of
```

Comment:
This might be simpler if you defined `max` and `min` on `Maybe Slot`? Except you probably can't do it on `Maybe Slot` since `Nothing` means something different depending which end it's at, maybe you need a stronger type.

Reply:
Yeah we could say `Interval a = Interval (Option (Min a)) (Option (Max a))` and then get all of that via the semigroup instance but we can't use type classes here :(